### PR TITLE
feat: group mentions + improved user mentions

### PR DESCRIPTION
**Pull request by alechkos**
# Table of Contents

### - [How Has the PR Been Tested](https://github.com/pedroslopez/whatsapp-web.js/pull/2602#tests) (latest test on 04.12.2023) <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2602#tests-b" id="tests-b"/>
- [**The environment the PR was tested on**](https://github.com/pedroslopez/whatsapp-web.js/pull/2602#env)

### - [Types of Changes](https://github.com/pedroslopez/whatsapp-web.js/pull/2602#types-of-changes) <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2602#types-of-changes-b" id="types-of-changes-b"/>

---

## [Description](https://github.com/pedroslopez/whatsapp-web.js/pull/2602#description-b) <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2602#description" id="description"/>
**This PR introduces logic for mentioning groups in messages and improves the logic for mentioning users:**
1. **Group mentions:**
    - Added option to send a message with clickable group mentions, and when the group mention is clicked, a chat with the mentioned group will be opened.
      > **NOTE:**
      > The user that does not participate in the mentioned group,
      > will not be able to open that mentioned group,
      > the same if the group does not exist
    - Added `Message.getGroupMentions` method to get an array of `GroupChat` objects from a message with group mentions
2. **User mentions:**
    - Added logic for mentioning those users with whom you have never interacted
    - Added a check to verify the existence of the user, preventing accidental mentions of unregistered phone numbers on WhatsApp

---

## [Usage Example](https://github.com/pedroslopez/whatsapp-web.js/pull/2602#usage-example-b) <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2602#usage-example" id="usage-example"/>

```js
client.on('message', async (msg) => {
    const chat = await msg.getChat();
    const userNumber = '26022602XX';

    if (msg.body === '!mentionUsers') {
        /**
         * To mention one user you can pass user's ID to 'mentions' property as is,
         * without wrapping it in Array, and a user's phone number to the message body:
         */
        await chat.sendMessage(`Hi @${userNumber}`, {
            mentions: userNumber + '@c.us'
        });

        // To mention a list of users:
        await chat.sendMessage(`Hi @${userNumber}, @${userNumber}`, {
            mentions: [userNumber + '@c.us', userNumber + '@c.us']
        });
    }
    
    if (msg.body === '!mentionGroups') {
        const chat = await msg.getChat();
        const groupId = 'YYYYYYYYYY@g.us';
        /**
         * Sends clickable group mentions, the same as user mentions.
         * When the mentions are clicked, it opens a chat with the mentioned group.
         * The 'groupMentions.subject' can be custom
         * 
         * @note The user that does not participate in the mentioned group,
         * will not be able to click on that mentioned group, the same if the group does not exist
         *
         * To mention one group:
         */
        await chat.sendMessage(`Check the last message here: @${groupId}`, {
            groupMentions: { subject: 'GroupSubject', id: groupId }
        });

        // To mention a list of groups:
        await chat.sendMessage(`Check the last message in these groups: @${groupId}, @${groupId}`, {
            groupMentions: [
                { subject: 'FirstGroup', id: groupId },
                { subject: 'SecondGroup', id: groupId }
            ]
        });
    }

    if (msg.body === '!getGroupMentions') {
        // To get group mentions from a message:
        const groupId = 'ZZZZZZZZZZ@g.us';
        const msg = await client.sendMessage('chatId', `Check the last message here: @${groupId}`, {
            groupMentions: { subject: 'GroupSubject', id: groupId }
        });

        /** {@link groupMentions} is an array of `GroupChat` */
        const groupMentions = await msg.getGroupMentions();
    }
});
```

---

## [To test this PR by yourself you can run one of the following commands:](https://github.com/pedroslopez/whatsapp-web.js/pull/2602#pr-testing-b) <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2602#pr-testing" id="pr-testing"/>

- **NPM**
```powershell
npm install github:alechkos/whatsapp-web.js#group-mentions
```
- **YARN**
```powershell
yarn add github:alechkos/whatsapp-web.js#group-mentions
```

---

## [If you encounter any errors while testing this PR, please provide in a comment:](https://github.com/pedroslopez/whatsapp-web.js/pull/2602#pr-issues-b) <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2602#pr-issues" id="pr-issues"/>

1. The code you've used without any sensitive information (use [syntax highlighting](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks#syntax-highlighting) for more readability)
2. The error you got
3. The library version
4. The WWeb version: `console.log(await client.getWWebVersion());`
5. The browser (Chrome/Chromium)

> [!IMPORTANT]
> **You have to [reapply](https://github.com/pedroslopez/whatsapp-web.js/pull/2602#pr-testing) the PR each time it is changed (new commits were pushed since your last application)**

---

## [How Has The PR Been Tested](https://github.com/pedroslopez/whatsapp-web.js/pull/2602#tests-b) (latest test on 04.12.2023) <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2602#tests" id="tests"/>
Tested with a code provided in usage example.

#### Types of accounts:
- Personal
- Buisness

#### Environment:
- Android 10:
    - WhatsApp: **latest**
    - WA Business: **latest**
- Windows 10:
    - WWebJS: **v1.23.1-alpha.0**
    - WWeb: **v2.2353.0**
    - Puppeteer: **v18.2.1**
    - Node: **v18.17.1**
    - Chrome: **latest**

---

## [Types of Changes](https://github.com/pedroslopez/whatsapp-web.js/pull/2602#types-of-changes-b) <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2602#types-of-changes" id="types-of-changes"/>
- [ ] Dependency change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix/feature that would cause existing functionality to change)

## Checklist
- [X] My code follows the code style of this project
- [X] I have updated the usage example accordingly (example.js)
- [X] I have updated the documentation accordingly (index.d.ts)

### CODE REVIEW

File: example.js [hunk 1]
```diff
@@ -237,12 +237,43 @@ client.on('message', async msg => {
...
         const chat = await msg.getChat();
-        chat.sendMessage(`Hi @${contact.number}!`, {
-            mentions: [contact]
+        // To mention one user you can pass the ID as is without wrapping it in Array:
+        await chat.sendMessage('Hi @number!', {
+            mentions: 'number@c.us'
         });
+        // To mention a list of users:
+        await chat.sendMessage('Hi @number1, @number2!', {
+            mentions: ['number1@c.us', 'number2@c.us']
```
Comment (shirser121) [inline example.js:248]: This is not example.
Please use:
```js
chat.sendMessage(`Hi @${contact.number}!`, {
            mentions: [contact]
        });
```

Comment (alechkos) [reply example.js:248]: In the last wweb versions you don't have to pass a contact object to mentions, you can simply pass a serialized id like this:
```js
clietn.sendMessage(groupId, 'test, @xxxxxxxxxxxxx', { mentions: ['xxxxxxxxxxxxx@c.us'] });
```

Comment (shirser121) [reply example.js:248]: I focus on the fact that 'number1@c.us' its not a real number

**Comment by crash-overdrive** [comment]
Using TypeScript, I can't use groupMentions as a MessageSendOption while calling the function client.SendMessage. Line 802-817 in Client.js should contain property groupMentions to fix this (I think).

**Comment by crash-overdrive** [comment]
> @crash-overdrive
> 
> > Line 802-817 in Client.js should contain property groupMentions to fix this
> 
> You are right, I forgot to add this Will be fixed

Lemme know when you plan on fixing this! Using it for a project of mine :D

File: index.d.ts [hunk 2]
```diff
@@ -850,7 +850,16 @@ declare namespace WAWebJS {
         /** MediaKey that represents the sticker 'ID' */
         mediaKey?: string,
         /** Indicates the mentions in the message body. */
-        mentionedIds: [],
+        mentionedIds: ChatId[],
```
[the remark this answers is not in the document]
Comment (alechkos) [reply index.d.ts:853]: This is not what we are passing to the `sendMessage` method, this is a structure of `mentionedIds` property of a `Message` object:
```js
const msg = await client.sendMessage(chatId, '@XXXXXXXXXX', { mentions: ['XXXXXXXXXX@c.us'] });
console.log(msg.mentionedIds); // HERE
```

[the remark this answers is not in the document]
Comment (crash-overdrive) [reply index.d.ts:853]: I agree, but if you console.log(msg.mentionedIds) it prints out just an array of string not an array of ChatId. ChatId has a few extra properties like server, id, _serialized (i am not sure what the properties are as i am not near my computer) while msg.mentionedIds just returns an array of serialized ids `XXXXXXXXXX@c.us`.

In typescript, msg.mentionedIds is no longer a string[] which it used to be. It’s a ChatId[] which somehow just prints _serialized property of ChatId if you console.log it.

**Review by matheus-de-araujo — APPROVED** [review]

**Comment by shavy4452** [comment]
is this ready for merge? I am waiting for this to include in my project.

**Review by tuyuribr — APPROVED** [review]

**Review by shirser121 — APPROVED** [review]
LGTM

**Pull request merged**